### DISCUSSION TITLE
Can I use Xrm.Data.PowerShell with Dynamics CRM 2013?

### DISCUSSION THREAD

**Discussion by h00jraq** [seanmcne/Microsoft.Xrm.Data.PowerShell#527] [Q&A] [answered]
Hi,

I'm investigating if there is any option to implement CI/CD pipelines for our Dynamics CRM 2013 version (hosted on-prem)
I see that, for example, Get-CRMconnection retrieves connection details for the specified Dataverse env or Dynamics 365.

Is it possible, with the help on those PS scripts, to for example export solution. extract it with Solution Packager, upload into GIT, work on some code there, package the solution with Package Manager again and import into CRM?

**Comment by seanmcne** [comment]
Get-CrmConnection is provided by the Microsoft provided xrm tooling module (not something authored here, though this module also relies on using that same cmdlet to get a connection object).  You have a couple of options, one would be to use Connect-CrmOnline but provide your own connectionstring using the connectionstring parameter (pointing to onprem).  

Second, you could just call the underlying [get-Connection](https://learn.microsoft.com/en-us/powershell/module/microsoft.xrm.tooling.crmconnector.powershell/get-crmconnection?view=pa-ps-latest) and use whatever parameters you want to create a CrmServiceClient. Once you have a connection created you can use all these helper cmdlets in this module to do what is needed. 

```
$conn = get-crmconnection {use parameters}

#get the whoamiresults - this will default to looking for the $conn variable 
Invoke-CrmWhoAmI

#get whoamiResults passing in the connection 
Invoke-CrmWhoAmI -conn $conn
```

**Reply by h00jraq** [reply, marked as answer]
I think the problem I face right now is the SDK which is used. I was trying to connect to Dynamics CRM 2015 while SDK being used is the newest SDK version (for Dynamics 365 as this is version 9?). I was trying to solve that problem but tbh I can't or I'm missing some info?

I can login normaly using the Url and it works fine with my credentials if that is important.

PS C:\Windows\system32> Get-CrmConnection -ConnectionString "AuthType=AD;Url=http://myserver:myport/xRMTraining/;Dom
ain=IHGD;UserName=myUserName;Password=password"
Get-CrmConnection : Failed to connect to CRM: Unable to login to Dynamics CRM, Error was : SOAP security negotiation
with 'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' for target
'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' failed. See inner exception for more
details.
SOAP security negotiation with 'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' for target
'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' failed. See inner exception for more
details. => The client credential entered was invalid.Unable to Login to Dynamics CRM
Unable to Login to Dynamics CRM
At line:1 char:1
+ Get-CrmConnection -ConnectionString "AuthType=AD;Url=http://myserver ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (:) [Get-CrmConnection], Exception
    + FullyQualifiedErrorId : -10,Microsoft.Xrm.Tooling.CrmConnector.Powershell.Commands.GetCrmConnectionCommand

Get-CrmConnection : Failed to connect to CRM: Unable to login to Dynamics CRM, Error was : SOAP security negotiation
with 'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' for target
'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' failed. See inner exception for more
details.
SOAP security negotiation with 'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' for target
'http://myserver:myport/xRMTraining/XRMServices/2011/Organization.svc' failed. See inner exception for more
details. => The client credential entered was invalid.Unable to Login to Dynamics CRM
Unable to Login to Dynamics CRM
At line:1 char:1
+ Get-CrmConnection -ConnectionString "AuthType=AD;Url=http://vmd1wxrmc ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : PermissionDenied: (:) [Get-CrmConnection], Exception
    + FullyQualifiedErrorId : -10,Microsoft.Xrm.Tooling.CrmConnector.Powershell.Commands.GetCrmConnectionCommand
    
    
I'm not sure what I'm doing wrong. Was trying to use interactive mode with "Connect-CrmOnPremDiscovery -InteractiveMode" and this one works and it returns below outpout:

IsReady                             : True
IsBatchOperationsAvailable          : True
MaxRetryCount                       : 10
RetryPauseTime                      : 00:00:05
Authority                           :
ActiveAuthenticationType            : AD
OAuthUserId                         :
TenantId                            : 00000000-0000-0000-0000-000000000000
EnvironmentId                       :
ConnectedOrgId                      : 943ca53a-0421-e511-9428-00505683690d
CrmConnectOrgUriActual              : http://myServer:myPort/xRMTraining/XRMServices/2011/Organization.svc
ConnectedOrgFriendlyName            : xRM Training
ConnectedOrgUniqueName              : xRMTraining
ConnectedOrgVersion                 : 7.0.0.3543
SdkVersionProperty                  : 9.2.47.9489
CallerId                            : 00000000-0000-0000-0000-000000000000
CallerAADObjectId                   :
DisableCrossThreadSafeties          : False
SessionTrackingId                   :
ForceServerMetadataCacheConsistency : False
LastCrmError                        :

I'm not sure how should I modify "Get-CrmConnection -ConnectionString in that case.

**Reply by h00jraq** [reply]
Ok I got it working after I used the value of CrmConnectOrgUriActual.
